### DISCUSSION TITLE
Add "armed_by" and "disarmed_by" attributes

### DISCUSSION THREAD

**Discussion by GizmoGoody** [tsightler/ring-mqtt#565] [Feature Requests]
Does the Ring API make the name of the User who last Armed/Disarmed the system available? If so, could those be added as attributes (e.g. armed_by and disarmed_by) under the keypad_info sensor?

Use case - Disarming the system is usually the first thing we do in the morning, but my wife and I prefer different scenes. It would be nice to be able to differentiate between who disarms the alarm for use in automations.

**Comment by tsightler** [comment]
There is an initiatingEntityId, which appears to be mapped to a user's name via a call to another API with that ID, which is how the name is retrieved and displayed in the Ring app/web console.  Unfortunately, I haven't been able to quite work out the exact details, but maybe this doesn't really matter as I've considering just forwarding the ID itself, since it wouldn't be very difficult to figure out which ID maps to what person.

**Comment by rbreneman** [comment]
+1 for this feature. I'd like to write an automation that would send a notification to the person who just armed the house (whether they used their code at keypad or used Ring App). I could also see use cases of automations for disarmed by as already mentioned. Thanks!

**Comment by tsightler** [comment]
Finally managed to work out how to do this.  There were two possible ways to go about it, try to grab the full arm/disarm event message as it happens, or query the event after the fact.  I decided to go for option one even though it was more complex because I thought it would delay the event notification less (event API can sometimes be a little bit slow or throttled).

Even grabbing the event update is a little bit tricky as there is still a delay as you have to query another API to get the user name.  This adds a small delay to the update, but my testing showed it being ~100ms which seemed acceptable.  The attributes are available on the alarm control panel entity and are as follows:

lastArmedBy
lastArmedTime
lastDisarmedBy
lastDisarmedTime

Feedback on this feature is greatly appreciated.  Not that there can be things other than users that trigger alarm mode changes, so, for example, if there is a mode schedule in the Ring app, this will display as "device-schedule" instead of a username.

**Reply by GizmoGoody** [reply]
Thanks! I’ll be away for another couple weeks, but should have time to play with this by the end of July.

**Reply by jasonabird** [reply]
This did work for me. I had to set my trigger to 1s to give it time for entity updates that if I put a condition on otherwise it wasn’t fully updated. Not a huge deal at all, just took some playing to realize that.

**Reply by tsightler** [reply]
Yeah, because the name is retrieved via a separate API call, it typically takes a couple hundred milliseconds, for the attribute data to be updated.  I had considered trying to delay the arm/disarm state update until the name had been retrieved, but I hate doing this because there could be weird cases where the websocket is still alive while the HTTPS API used for name lookups is not, so the call to the API may never return or take a very long time to timeout.

In general, I lean towards providing state updates as soon as I have them, so I decided to just follow that logic in this case as well, but it could perhaps be improved.  Maybe I should implement a wait up to one second for the name to resolve before sending the state update then they would be more in sync.

**Comment by jasonabird** [comment]
Was just using this in an automation and found that Guest Users do not show the name of the user you put in Ring, just come through as “user”. Assuming the ring api shows the name of the guest user? Would be great if that would add in the entry the name vs “user” for who armed/disarmed.

**Reply by tsightler** [reply]
Why would you assume that? 
 Actually, the API does not provide a name at all.  It provides an entity type and, if the disarming entity type is user it provides an entity id that you then have to call another API to lookup.  If this lookup fails then then only thing known is that it was a user that disarmed it vs, for example, a schedule, so that is what ring-mqtt reports.

What do guest users show up in the Ring app?  If it somehow reports guest user names feel free to reverse engineer how it does it and submit a PR.

**Reply by jasonabird** [reply]
Sounds like a completely invalid assumption then on my part. Didn’t realize you had to do additional entity lookups to get the name. The guest users in the ring app/notifications shows the name of the guest user.  Haven’t had chance to fully reverse engineer how it works at all.

**Reply by tsightler** [reply]
I will try to take a look at it at some point, perhaps there is another API call needed to get the names for guest users vs regular users.  I honestly just assumed the same API call would work for both user types, so that's me perhaps making an invalid assumption.

**Reply by tsightler** [reply]
So I think I finally figured this out.  For primary/shared users the event includes only the entity ID, but for guest users, the event includes a different property that actually does have the name, so, for regular users, an additional lookup is required, but for guest users it is not.  Hopefully it will work in the next released version.

**Reply by jasonabird** [reply]
Great to hear!

**Comment by dom8925** [comment]
Was this every implemented? Just started using this and am not seeing my name when I arm/disarm

**Reply by tsightler** [reply]
Yes, armedBy And disarmedBy attributes are populated when possible.  Some actions, like disarming by a schedule, might not show anything.

**Reply by dom8925** [reply]
I disarmed on my key pad but HA only shows that it was disarmed. I am the owner. 
![SmartSelect_20240727_160844_Home Assistant](https://github.com/user-attachments/assets/d5f8d5be-698e-488c-b4d3-0371adafab1e)
Do I need to configure something to make it show my name?

**Reply by tsightler** [reply]
Ring-mqtt has zero control over what HA logs.  The data is included as an attribute on the alarm control panel (you can see it if you select Developer Tools and select your alarm entity) and this is consumable via automation or UI element so you can do with it whatever you choose to do from an automation, notification or UI perspective.

**Reply by dom8925** [reply]
Got it. That's what I was missing. I see it now. Thanks!
